### PR TITLE
[doc] update advanced tutorials, training gpt with hybrid parallelism

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4865 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

update advanced tutorials, training gpt with hybrid parallelism

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md [hunk 1]
```diff
@@ -1,271 +1,203 @@
...
+import colossalai
+from colossalai.booster import Booster
+from colossalai.booster.plugin import GeminiPlugin, HybridParallelPlugin, LowLevelZeroPlugin, TorchDDPPlugin
+from colossalai.cluster import DistCoordinator
+from colossalai.nn.optimizer import HybridAdam
+from colossalai.utils import get_current_device
+```
 
-For GPT, the *word embedding layer* shares the weights with the *output head*. We provide `PipelineSharedModuleWrapper` to share parameters among pipeline stages. It takes a `list` of `int` as argument, which means those ranks share the parameters. You can use `register_module()` or `register_parameter()` to register a module or a parameter as the shared module or parameter. If you have multiple sets of shared modules / parameters, you should have multiple `PipelineSharedModuleWrapper` instance. If the parameter is shared within **one** stage, you should not use `PipelineSharedModuleWrapper`, and just use the same module / parameter instance. In this example, the *word embedding layer* is at the first stage, and the *output head* is at the last stage. Thus, they are shared among ranks `[0, pipeline_size - 1]`.
+## Define GPT-2‘s Training Components
```
Comment (ver217) [inline docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md:54]: Use `'` rather than `’`.

File: docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md [hunk 2]
```diff
@@ -1,271 +1,203 @@
...
-    model = GPT2_exlarge_pipeline_hybrid(num_chunks=num_chunks, checkpoint=True, dtype=torch.half)
-    # model = GPT3_pipeline_hybrid(num_chunks=num_chunks, checkpoint=True, dtype=torch.half)
-    if use_interleaved and not isinstance(model, nn.ModuleList):
-        model = nn.ModuleList([model])
-
-    criterion = GPTLMLoss()
-
-    optimizer = torch.optim.Adam(model.parameters(), lr=0.00015, weight_decay=1e-2,)
+data_builder = GLUEDataBuilder(
+    model_name, plugin, args.task, train_batch_size=BATCH_SIZE, eval_batch_size=BATCH_SIZE
```
Comment (ver217) [inline docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md:75]: Where do you define `model_name`?

File: docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md [hunk 3]
```diff
@@ -1,271 +1,203 @@
...
-        return_output_label=False,
-    )
 ```
-<!-- doc-test-command: echo  -->
+Training the gpt-2 model
+```python
+for epoch in range(NUM_EPOCHS):
+    train_epoch(epoch, model, optimizer, _criterion, lr_scheduler, train_dataloader, booster, coordinator)
+```
+<!-- doc-test-command: echo  -->
```
Comment (ver217) [inline docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md:203]: Why not run the test?

File: docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md [hunk 4]
```diff
@@ -1,271 +1,211 @@
...
+    is_pp_last_stage = use_pipeline and booster.plugin.stage_manager.is_last_stage()
+    total_step = len(train_dataloader)
+
+    model.train()
+    optimizer.zero_grad()
+    train_dataloader_iter = iter(train_dataloader)
+    with tqdm(
+        range(total_step),
+        desc=f"Epoch [{epoch + 1}/{NUM_EPOCHS}]",
+        disable=not (coordinator.is_master() or is_pp_last_stage),
```
Comment (ver217) [inline docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md:180]: This condition may wrong

File: docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md [hunk 5]
```diff
@@ -1,271 +1,211 @@
...
+
+def _criterion(outputs, inputs):
+    outputs = output_transform_fn(outputs)
+    loss = criterion(outputs)
+    return loss
+```
+## Boost the GPT-2 Model
+Define a booster with `HybridParallelPlugin`. Based on the configured plugin parameters, the booster will inject one or more parallel strategies into the model. In this example, pipeline parallelism, zero1, and mixed-precision training optimizations are utilized.
+```python
+booster_kwargs=dict(mixed_precision='fp16')
```
Comment (ver217) [inline docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md:144]: Precision is specified in plugin configuration. I think you can remove this, right?

File: docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md [hunk 6]
```diff
@@ -1,271 +1,211 @@
...
+    {
+        "params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
+        "weight_decay": 0.0,
+    },
+]
+optimizer = HybridAdam(optimizer_grouped_parameters, lr=lr, eps=1e-8)
+```
+Prepare the lr_scheduler and criterion, and it's important to note that when hybrid parallelism with pipeline parallelism is used, a criterion function should also be defined. This function should take the input and output of the model's forward pass as parameters and return the loss.
+```python
+output_transform_fn = lambda x: x
```
Comment (ver217) [inline docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md:125]: Is seems that `output_transform_fn` and `criterion` are useless. They are only use in `_criterion`. Why not move these codes to `_criterion`?

File: docs/source/en/advanced_tutorials/train_vit_using_pipeline_parallelism.md [hunk 7]
```diff
@@ -2,8 +2,12 @@
 
 Author: Hongxin Liu, Yongbin Li
 
+**Prerequisite:**
```
Comment (ver217) [inline docs/source/en/advanced_tutorials/train_vit_using_pipeline_parallelism.md:5]: This tutorial is similar to vit using hybrid parallelism. Why not remove this one?

File: docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md (hunk 5 above)
Comment (flybird11111) [reply docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md]: updated.

File: docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md (hunk 6 above)
Comment (flybird11111) [reply docs/source/en/advanced_tutorials/train_gpt_using_hybrid_parallelism.md]: updated

File: docs/source/en/advanced_tutorials/train_vit_using_pipeline_parallelism.md (hunk 7 above)
Comment (flybird11111) [reply docs/source/en/advanced_tutorials/train_vit_using_pipeline_parallelism.md]: updated

**Review by ver217 — APPROVED** [review]

**Review by FrankLeeeee — APPROVED** [review]
